mlflow/pipelines/utils/execution.py

```diff
@@ -281,6 +305,21 @@ def _run_make(execution_directory_path, rule_name: str, extra_env: Dict[str, str
     :param extra_env: Extra environment variables to be defined when running the Make child process.
     :param rule_name: The name of the Make rule to run.
     """
+    # Dry-run Make with verbose debugging outputs, filter the outputs to only show which cached
+    # data targets are used.
+    process = _exec_cmd(
+        ["make", "--debug=v", "-n", "-f", "Makefile", rule_name],
```

Comment:
Rather than printing and parsing the verbose make `--debug` trace, whose contents may vary depending on the platform, it seems safer to add comments to the Makefile command for each step and then filter the stdout for the predefined comment string,

 e.g. add the following immediately above   https://github.com/mlflow/mlflow/blob/72169d76095272a5ff213aabdff4e088f993f6cb/mlflow/pipelines/utils/execution.py#L416

```
# Running MLP step: ingest
```

and add the following immediately above https://github.com/mlflow/mlflow/blob/72169d76095272a5ff213aabdff4e088f993f6cb/mlflow/pipelines/utils/execution.py#L423

```
# Running MLP step: train
```

and so on. Then, parse the output of `make -n` for strings of the form "# Running MLP step: *"

Reply:
In this case, we will know which steps will be executed before actually run the Make. For example, for the following cmd:
```
> mlp run -p local -s train
```
suppose this call uses cached steps `ingest` and `split`, and will run `transform` and `train`. After parsing the output of `make -n`, we can print the following lines:
```
1. Skipping cached MLP step: ingest
2. Skipping cached MLP step: split
3. Running MLP step: transform
4. Running MLP step: train 
```
Then we call `make` to actually run the steps, which generates the following output lines:
```
5. [outputs when running transform]
6. [outputs when running train]
```
Would we want the output to be "1,2,3, **5,4,** 6" instead of "1,2,3,4,5,6"?

Reply:
Discussed offline:

For Q1, we want to output 1 (cached), 2 (cached), 3 running ... output, 4 running ... output

For Q2: I think inferring the missing steps is a good approach, since MLP has to know the structure of the DAG anyway to generate the makefile in the future. Right now, we rely on a particular DAG structure in other parts of the execution logic, e.g. https://github.com/mlflow/mlflow/blob/14088b43b909e27b260ba132868699587a07be22/mlflow/pipelines/utils/execution.py#L82-L90, which assumes an ordered list of steps.

Reply:
This PR https://github.com/mlflow/mlflow/pull/6455 breaks the assumption you mentioned above.
It added a **disjoint** DAG (inference pipeline) while the `_BasePipeline._steps` and execution.py https://github.com/mlflow/mlflow/blob/14088b43b909e27b260ba132868699587a07be22/mlflow/pipelines/utils/execution.py#L82-L90 assumes the pipeline contains **one connected DAG**.

Furthermore, we fix the logic to reload `self._steps` from the config files. Right now, it reloads at 3 places:
1. In `_BasePipeline.__init__()`.
2. Before each execution: https://github.com/mlflow/mlflow/blob/ff874bc935a923c4eafa9ff091145fa0492c8c13/mlflow/pipelines/pipeline.py#L82
3. in `_BasePipeline._get_step(...)` https://github.com/mlflow/mlflow/blob/ff874bc935a923c4eafa9ff091145fa0492c8c13/mlflow/pipelines/pipeline.py#L147 which is called in `run()`, `clean()` and `inspect()`

This is a bit hard to reason about when can we trust the `self._steps` loaded in the pipeline and when should we reload the steps from the config file.

**Action items:**
1. _BasePipeline should be aware of 2 ordered list of steps. One is `_training_steps`, another is `_scoring_steps`. 
2. We should only reload `self.*_steps` in the beginning of `__init__()`, `run()`, `clean()` and `inspect()`, and assume they contain up-to-date info elsewhere.
3. Fix this issue in a **dedicated PR** and resolve this TODO before implementing the cache output.
https://github.com/mlflow/mlflow/blob/ff874bc935a923c4eafa9ff091145fa0492c8c13/mlflow/pipelines/pipeline.py#L88
We shouldn't add new feature to incorrect code path.

cc @dbczumar @jinzhang21

Reply:
@liangz1 I agree with your assessment of the problem, but I think it's sufficient to reload all of the steps (rather than just the subgraph steps) when `__init__()`, `run()`, `clean()`, and `inspect()` are called. `Pipeline` doesn't have to know about scoped steps to achieve a correct solution, right?

Reply:
> Pipeline doesn't have to know about scoped steps to achieve a correct solution, right?

Right. We can allow Pipeline to be unaware of the subgraph. Then we need to generate (filter out) the subgraph according to `_TRAIN_DAG_STEPS, _SCORE_DAG_STEPS`. The assumption of `_BasePipeline._steps` is: It contains concatenated ordered lists representing multiple disjoint DAGs. So the current order of `_steps` is also valid:
`ingest_scoring, predict, ingest, split, transform, train, evaluate, register`.

Reply:
Great! Can we make that change to `RegressionPipeline` as part of this PR? I don't think this work is so complex that it requires a separate PR, right?

Reply:
I would prefer it to be done in a separate PR since it's logically independent to this feature. We also need to add tests for the fix. I think mixing the fix & new feature will make the PR difficult to review.

Reply:
Okay :)